core/arch/arm/plat-zynqmp/drivers/aes.c

```diff
@@ -0,0 +1,295 @@
+	if (ret) {
+		EMSG("DMA tag transfer timeout");
+		return ret;
+	}
+
+	/* Wait for completion*/
+	ret = csu_dma_sync(CSU_DMA_DST_CHANNEL);
+	if (ret) {
+		EMSG("DMA destination transfer timeout");
+		return ret;
+	}
+
+	csu_dma_unprepare();
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result aes_prepare_op(enum aes_op op, enum csu_key key)
+{
+	vaddr_t aes = core_mmu_get_va(CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      CSU_AES_SIZE);
+	vaddr_t csu = core_mmu_get_va(CSU_BASE, MEM_AREA_IO_SEC,
+				      0x1000);
+	TEE_Result ret = TEE_SUCCESS;
+
+	/* Connect DMA0 in/out to AES */
+	io_write32(csu + SSS_OFFSET, DMA0_STREAM_TO_AES);
+
+	/* Reset the AES */
+	io_write32(aes + AES_RESET_OFFSET, AES_RESET_SET);
+	io_write32(aes + AES_RESET_OFFSET, AES_RESET_CLR);
+
+	/* Load the key */
+	io_write32(aes + AES_KEY_CLR_OFFSET, 0);
+	io_write32(aes + AES_KEY_SRC_OFFSET, key);
+	io_write32(aes + AES_KEY_LOAD_OFFSET, AES_KEY_LOAD);
+	ret = aes_wait(AES_STS_KEY_INIT_DONE, true);
+	if (ret) {
+		EMSG("Timeout loading the key");
+		return TEE_ERROR_GENERIC;
+	}
+
+	/* Configure operation */
+	io_write32(aes + AES_CFG_OFFSET,
+		   op == AES_DEC ? AES_CFG_DEC : AES_CFG_ENC);
+
+	/* Prepare the CSU for the DMA */
+	io_write32(csu + DMA_RESET_OFFSET, DMA_RESET_SET);
+	io_write32(csu + DMA_RESET_OFFSET, DMA_RESET_CLR);
+
+	/* Start the message */
+	io_write32(aes + AES_START_MSG_OFFSET, AES_START_MSG);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result aes_done_op(enum aes_op op, TEE_Result ret)
+{
+	vaddr_t aes = core_mmu_get_va(CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      CSU_AES_SIZE);
+	uint32_t val = 0;
+
+	/* Validate the GCM tag */
+	if (!ret && op == AES_DEC) {
+		val = io_read32(aes + AES_STS_OFFSET) & AES_STS_GCM_TAG_OK;
+		if (!val) {
+			EMSG("AES-GCM tag mismatch");
+			return TEE_ERROR_GENERIC;
+		}
+	}
+
+	val = io_read32(aes + AES_KEY_CLR_OFFSET);
+	io_write32(aes + AES_KEY_CLR_OFFSET, val | AES_KEY_ZERO | AES_KUP_ZERO);
+	if (aes_wait(AES_STS_AES_KEY_ZEROED | AES_STS_KUP_ZEROED, true))
+		EMSG("Failed to clear the AES key");
+	io_write32(aes + AES_KEY_CLR_OFFSET, val);
+
+	io_write32(aes + AES_RESET_OFFSET, AES_RESET_SET);
+
+	if (ret)
+		return ret;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result aes_decrypt_data(uint8_t *src, uint8_t *dst, uint8_t *tag,
+			    uint8_t *iv, size_t len, enum csu_key key)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (key != CSU_AES_KEY_SRC_DEV) {
+		EMSG("Key type not supported");
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	if (len % 4) {
+		EMSG("Invalid source size");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = aes_prepare_op(AES_DEC, key);
+	if (ret) {
+		EMSG("Decrypt init failed");
+		goto error;
+	}
+
+	ret = aes_transfer_dec(src, dst, tag, iv, len);
+	if (ret) {
+		EMSG("DMA transfer failed");
+		goto error;
+	}
+
+	ret = aes_wait(AES_STS_AES_BUSY, false);
+	if (ret)
+		EMSG("AES-GCM transfer failed");
+error:
+	return aes_done_op(AES_DEC, ret);
+}
+
+TEE_Result aes_encrypt_data(uint8_t *src, uint8_t *dst, uint8_t *iv,
```

Comment:
Often order of buffers and their sizes are next to each other.

Like `src, src_len, dst, dst_len...`

And should these pointer types be `void`?

`src` and `iv` with const.

Reply:
yeah you are right